### PR TITLE
Breaking: Switch to ESM

### LINKED ISSUE #22: Convert package to ESM
To convert this package from CommonJS to ECMAScript modules (ESM):

- [ ] Add `type: module` to `package.json`
- [ ] Convert files in `lib` to ESM
- [ ] Convert files in `tests/lib` to ESM
- [ ] Upgrade Mocha
- [ ] Create `build` script in `package.json` to generate CommonJS file (`.cjs`) in `dist` directory (use Rollup)
- [ ] Add `dist` to `.gitignore`
- [ ] Run `build` in `prepare` script in `package.json` to ensure build is run at key moments
- [ ] Update `package.json` so that `main` points in the CJS file in `dist` and `exports` provides both `require` and `import` options
- [ ] Update README instructions

### PR DESCRIPTION
**Pull request by brettz9**
Fixes #22.

Also:

- Refactor(lint): Fix `Object.prototype.hasOwnProperty` call as per latest linting requirements
- Chore(lint): Adjust `.eslintrc.json` to pass linting of modules; uses ES2021 to support `import.meta.url`
- Chore(editorconfig): Add `.editorconfig`
- Chore(npm): Update devDeps.
- Chore(npm) Switches from nyc to c8 due to nyc requiring extra loader: https://github.com/istanbuljs/nyc/issues/1287

Using synchronous fs API since top-level await not yet supported (nor is ESM importing of JSON well-supported).

(I'm not sure why in my IDE, Atom, I'm seeing `node/no-unsupported-features/es-syntax` errors but in the `lint` script I am not.)

### CODE REVIEW

**Comment by brettz9** [comment]
If I can get feedback about whether I'm on the right track with this, I may be able to also look at https://github.com/eslint/eslint-scope/issues/70 and https://github.com/eslint/eslintrc/issues/35 .

**Review by nzakas — DISMISSED** [review]
This LGTM. I’d like another set of eyes on it before merging. 

Thanks for doing this!

File: package.json
Comment (aladdin-add) [inline package.json:28]: what is the `-c` used for?

File: package.json
Comment (aladdin-add) [reply package.json:9]: should add './dist' to the "files", or it will break cjs users.

**Review by aladdin-add — CHANGES_REQUESTED** [review]

File: package.json
Comment (brettz9) [reply package.json:28]: If you don't supply the `-c`, the command just gives a help screen. If you put something after the flag, you can indicate a specific config file, but otherwise it defaults to `rollup.config.js` which is what I've done here.

File: package.json
Comment (brettz9) [reply package.json]: Sure. To make it easier to merge, I've added a commit to mirror your current engines changes but can of course update later according to any subsequent changes.

File: package.json
Comment (brettz9) [reply package.json]: Ah, sorry, yes. I've added `dist` (and also `package.json`).

Should we perhaps also export `lib/visitor-keys.json`?

File: package.json
Comment (brettz9) [reply package.json:34]: Yes, `nyc` currently does not work in a stable way with ESM, while `c8` does.

While the `coverage` script change is not shown in the diff there, it too is still present but uses `c8` instead.

File: package.json
Comment (brettz9) [reply package.json]: By the way, I've added subsequent commits with the expected commit messages, e.g., "Fix:", etc. so as to avoid wiping out the commit history. Not sure if you prefer my amending instead, though in any case, I can of course squash it later as deisred.

Comment (aladdin-add) [reply package.json]: no worries, we will squash the commits when merging. 😄

File: package.json
Comment (aladdin-add) [reply package.json:32]: well, I see! I just thought it will auto load the config like eslint. :)

File: package.json
Comment (aladdin-add) [reply package.json]: IMO, no need to export the `lib/visitor-keys.json`, as it was just an internal module. e.g. we may rename it in future releases.

**Review by aladdin-add — DISMISSED** [review]
LGTM, thanks!

**Comment by mdjermanovic** [comment]
We already have one ESM project: https://github.com/eslint/espree. Ideally, all repositories should be aligned with espree repo in that regard.

File: README.md
Comment (mdjermanovic) [inline README.md:25]: We should have both in the docs, like in https://github.com/eslint/espree#usage

File: package.json
Comment (mdjermanovic) [inline package.json:6]:
```suggestion
  "main": "dist/eslint-visitor-keys.cjs",
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: README.md
Comment (brettz9) [reply README.md]: Sure, added.

File: lib/index.js
Comment (mdjermanovic) [inline lib/index.js:8]: This doesn't work in `dist/eslint-visitor-keys.cjs`, it tries to load `dist/visitor-keys.json`.

Also, `eslint-visitor-keys` is used by `espree` and ESLint's `Linter`, which are intended to work in browsers as well (e.g., in our [Online Demo](https://eslint.org/demo)). Even if we manage to fix the issue with `dist/eslint-visitor-keys.cjs`, I'm not sure if bundlers used in projects that depend on `eslint-visitor-keys` will be able to properly handle this.

Thoughts about converting `visitor-keys.json` to `visitor-keys.js` module? That would avoid any possible issues with loading local files.

**Comment by brettz9** [comment]
@mdjermanovic : Do the issues you mentioned comprise the alignment changes you still feel or needed, or were there any other dimensions where you wanted alignment? I've submitted some of the other changes I aaw.

Note that I didn't align with nyc because that is broken with ESM, and I didn't align the rollup script in JSON to specify `rollup.config.js` as that seems redundant.

Btw, you are using `prepublishOnly` to invoke the build script while @nzakas mentioned a `prepare` script. Not sure which you want.

**Comment by brettz9** [comment]
Btw, per my alignment with espree, I didn't see much mention of the array you used with `exports`, though I do see mention of a fallback array in https://nodejs.org/api/packages.html (though nothing else). But I'll take your word it works. :-)

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: lib/index.js
Comment (brettz9) [reply lib/index.js:8]: SGTM, esp. if there is deemed to be no need to export that JSON file as discussed previously.

**Comment by mdjermanovic** [comment]
> @mdjermanovic : Do the issues you mentioned comprise the alignment changes you still feel or needed, or were there any other dimensions where you wanted alignment? I've submitted some of the other changes I aaw.

Thanks for the changes! I'll check everything again and let you know if there is anything left.

> Note that I didn't align with nyc because that is broken with ESM

That's great, coverage is currently broken in the espree repo.

> and I didn't align the rollup script in JSON to specify `rollup.config.js` as that seems redundant.

:+1:  that's fine 

> Btw, you are using `prepublishOnly` to invoke the build script while @nzakas mentioned a `prepare` script. Not sure which you want

Espree had an additional complexity with [`update-version`](https://github.com/eslint/espree/blob/c8ca49a281eb74bb81e9bc3a60300011e8b29e8d/package.json#L76). `prepare` should be fine for this project, and it's also simpler and better for various reasons.

> Btw, per my alignment with espree, I didn't see much mention of the array you used with exports

That was discussed in the [RFC](https://github.com/eslint/rfcs/pull/72). It's for compatibility with some Node versions, though I'm not sure if we need that anymore after https://github.com/eslint/eslint-visitor-keys/pull/23.

**Comment by brettz9** [comment]
Ok, cool.

As far as the fallback array, since it seems less documented and more verbose, maybe it is better to avoid it if the version support mentioned in https://github.com/eslint/rfcs/pull/72#issuecomment-759171619 are accurate. Let me know if you want to keep it or revert (and just add the "default" key).

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:8]: `visitor-keys.json` wasn't documented in README, so I think it's fine to remove it in a major version.

@nzakas thoughts about this?

File: lib/index.js
Comment (mdjermanovic) [inline lib/index.js:40]: Would named exports make more sense than `export default`, since `KEYS`, `getKeys()` and `unionWith()` look more like separate features than properties of an object?

For example, typescript-eslint imports this library with [`import * as eslintVisitorKeys from 'eslint-visitor-keys'`](https://github.com/typescript-eslint/typescript-eslint/blob/8cfe93372e1d826e54febc3aeb7047c792b90963/packages/visitor-keys/src/visitor-keys.ts#L2).

Comment (brettz9) [reply lib/index.js:40]: Yeah, this is an issue to decide on the other repos needing changing as well. I think separate is good to allow for tree-shaking, but y'all's call. Let me know.

File: lib/index.js
Comment (nzakas) [reply lib/index.js:8]: Agree

**Comment by nzakas** [comment]
Regarding prepare vs prepublishOnly: in general we want to use prepare everywhere. Espree had a special case where that wouldn’t work.

File: lib/index.js
Comment (brettz9) [reply lib/index.js]: Ok, switched to a module.

**Comment by mdjermanovic** [comment]
Linting fails due to missing plugins required by the new `eslint-config-eslint`. You can add `eslint-plugin-node` and `eslint-plugin-jsdoc`.

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:36]: Let's do named exports, without `export default`.

**Comment by mdjermanovic** [comment]
It would be nice to have a small test file that would `require("../../dist/eslint-visitor-keys.cjs")`, and then check e.g. if `getKeys` is a function etc., similar to this one:

https://github.com/eslint/espree/blob/main/tests/lib/commonjs.cjs

File: rollup.config.js
Comment (mdjermanovic) [inline rollup.config.js:3]:
```suggestion
```

This can be removed now?

Comment (brettz9) [reply rollup.config.js:3]: Right, removed.

File: .eslintrc.json
Comment (mdjermanovic) [inline .eslintrc.json:12]: This is now a duplicate `"env"` key. I think we can just remove this one, the other `"env"` enables `es2020`, and `eslint-plugin-node` enables appropriate globals for esm and cjs files.

Comment (brettz9) [reply .eslintrc.json]: Sorry, missed that. Done.

**Comment by brettz9** [comment]
> It would be nice to have a small test file that would `require("../../dist/eslint-visitor-keys.cjs")`, and then check e.g. if `getKeys` is a function etc., similar to this one:
> 
> https://github.com/eslint/espree/blob/main/tests/lib/commonjs.cjs

Added.

File: lib/index.js
Comment (mdjermanovic) [reply lib/index.js:36]: > Let's do named exports, without `export default`.

To be clear, this is only about `lib/index.js`. 

`export default` is okay in `lib/visitor-keys.js`

Comment (brettz9) [reply lib/index.js:36]: Do you want the exports at the end of the file or inline?

Comment (mdjermanovic) [reply lib/index.js:36]: Maybe inline, like in espree?

**Comment by mdjermanovic** [comment]
It might be more suitable to do this freezing in `visitor-keys.js` now? 

https://github.com/eslint/eslint-visitor-keys/blob/7279e7304e95030a854408191b8fde3c01876451/lib/index.js#L12-L16

File: lib/index.js
Comment (brettz9) [reply lib/index.js]: Ok, done.

File: README.md
Comment (mdjermanovic) [inline README.md:27]:
```suggestion
import * as evk from "eslint-visitor-keys"
```

**Comment by brettz9** [comment]
> It might be more suitable to do this freezing in `visitor-keys.js` now?
> 
> https://github.com/eslint/eslint-visitor-keys/blob/7279e7304e95030a854408191b8fde3c01876451/lib/index.js#L12-L16

Done.

File: package.json
Comment (mdjermanovic) [inline package.json:19]: Do we want to publish `eslint-visitor-keys.cjs.map` as well, or only `eslint-visitor-keys.cjs`?

File: .eslintrc.json
Comment (mdjermanovic) [inline .eslintrc.json:8]:
```suggestion
        "ecmaVersion": 2020
```

File: .eslintrc.json
Comment (mdjermanovic) [inline .eslintrc.json:17]:
```suggestion
    ],
    "ignorePatterns": "/dist"
```

File: package.json
Comment (mdjermanovic) [inline package.json:39]:
```suggestion
    "lint": "eslint .",
```

to lint `rollup.config.js` and `commonjs.cjs`, too.

**Comment by mdjermanovic** [comment]
Does `npm run coverage` works for you? I'm getting some errors.

**Comment by brettz9** [comment]
Looks like it is this bug recurring: https://github.com/bcoe/c8/issues/184 . 

(Logging shows that there is a "branchMap" expected but it only has keys 0-6, not key 7 as supplied to it from "b" as obtained from "branches".)

**Comment by mdjermanovic** [comment]
Hm, it works when I remove `commonjs.cjs` test file. Could we maybe split `*.js` and `*.cjs` test runs and run only `*.js` tests through `c8`, like in the espree repo?

**Comment by brettz9** [comment]
Added a commit to split the tests

**Comment by mdjermanovic** [comment]
> Added a commit to split the tests

It works now, thanks!

File: package.json
Comment (mdjermanovic) [reply package.json:19]:
```suggestion
    "dist/eslint-visitor-keys.cjs",
```

Let's publish only the module for now, like with espree. If someone asks for the map file, we can add it later.

File: lib/visitor-keys.js
Comment (nzakas) [inline lib/visitor-keys.js:292]: Did you mean to freeze an array here? I don't think this is necessary because we aren't exporting `NODE_TYPES`.

File: package.json
Comment (nzakas) [inline package.json:12]: I'm pretty sure we don't need both "default" and "require" here, one of them should suffice.

File: package.json
Comment (nzakas) [reply package.json:19]: Agreed, we don't need `.map`.

File: lib/visitor-keys.js
Comment (brettz9) [reply lib/visitor-keys.js]: I've stopped freezing it. I just copied it over from the index page where it also wasn't been used, so yeah, can be removed.

**Comment by brettz9** [comment]
Regarding the `dist` not exporting map, I've made the change, but just to be clear, as per the parity with espree, the Rollup file does build a sourcemap, so it is available for the repo/development at least.

Likewise with `default` having been carried over.

File: package.json
Comment (brettz9) [reply package.json]: I went ahead and removed, but note espree is still using it (the reason why I added it previously). I imagine a bundler could conceivably treat it differently, though Rollup is always satisfied if at least `import` or `require` is present. I don't see from https://webpack.js.org/guides/package-exports/ whether Webpack does so, but I would suspect it should work.

This leaves:

```json
"exports": {
    ".": [
      {
        "import": "./lib/index.js",
        "require": "./dist/eslint-visitor-keys.cjs"
      },
      "./dist/eslint-visitor-keys.cjs"
    ],
    "./package.json": "./package.json"
  },
```

...though I think we could also forego the fallback array if you want.

**Review by nzakas — APPROVED** [review]

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

File: package.json
Comment (nzakas) [reply package.json]: Ah, I missed that Espree was using it. I'm pretty sure "default" and the fallback array are redundant with "require" in all relevant cases, but I'm fine with leaving this as-is.

**Pull request merged**
